File: Sudoku/cpp/sdku.cpp
// Sudoku solution in C++ for CodeEval.com by Steven A. Dunn

#include <algorithm>
#include <cmath>
#include <cstdlib>
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

using std::cout;
using std::endl;
using std::ifstream;
using std::istringstream;
using std::replace;
using std::string;
using std::vector;

vector<int> parseLine(string);
string isValid(int, vector<int>);
bool rowCheck(int, vector<int>);
bool colCheck(int, vector<int>);
bool gridCheck(int, vector<int>);
int getSumForCheck(int);

int main(int argc, char* argv[])
{
  ifstream inputFile(argv[1]);
  string line;

  if (inputFile)
  {
    while (getline(inputFile, line))
    {
      int N = line.at(0) - '0';
      vector<int> nums = parseLine(line);
      cout << isValid(N, nums) << endl;
    }
    inputFile.close();
  }
  return 0;
}

vector<int> parseLine(string line)
{
  vector<int> results;
  line = line.substr(line.find(";") + 1);
  replace(line.begin(), line.end(), ',', ' ');
  istringstream iss(line);
  string s;
  while (iss >> s)
    results.push_back(atoi(s.c_str()));
  return results;
}

string isValid(int N, vector<int> nums)
{
  if (rowCheck(N, nums) && colCheck(N, nums) && gridCheck(N, nums))
    return "True";
  return "False";  
}

bool rowCheck(int N, vector<int> nums)
{ 
  int sumForCheck = getSumForCheck(N);

  for (int i = 0; i < N * N; i += N)
  {
    int rowSum = 0;
    for (int j = 0; j < N; ++j)
      rowSum += nums[i + j];

    if (rowSum != sumForCheck)
      return false;
  }
  return true;
}

bool colCheck(int N, vector<int> nums)
{
  int sumForCheck = getSumForCheck(N);

  for (int i = 0; i < N; ++i)
  {
    int colSum = 0;
    for (int j = 0; j < N * N; j += N)
      colSum += nums[i + j];
  
    if (colSum != sumForCheck)
      return false;
  }

  return true;
}

bool gridCheck(int N, vector<int> nums)
{
  int sumForCheck = getSumForCheck(N);
  int sqrtN = sqrt(N);
  for (int gridX = 0; gridX < sqrtN; ++gridX) 
  {
    for (int gridY = 0; gridY < sqrtN; ++gridY)
    {
      int gridSum = 0;
      for (int i = 0; i < sqrtN; ++i)
        for (int j = 0; j < sqrtN; ++j)
          gridSum += nums[N * i + j + gridY * (N * sqrtN) + gridX * (sqrtN)];
      if (gridSum != sumForCheck)
        return false;
    }
  }
  
  return true;
}

int getSumForCheck(int N)
{
  int result;
  if (N == 4)
    result = 4 + 3 + 2 + 1;
  else
    result = 9 + 8 + 7 + 6 + 5 + 4 + 3 + 2 + 1;
  return result;
}

File: Sudoku/py2/sdku.py
# Sudoku solution in Python 2 for CodeEval.com by Steven A. Dunn

import math, sys

def get_sum_for_check(N):
  if N is 4:
    return 10
  else:
    return 45

def row_check(N, numbers):
  sum_for_check = get_sum_for_check(N)
  for i in range(0, N * N, N):
    row_sum = 0
    for j in range(0, N):
      row_sum += numbers[i + j]

    if not row_sum is sum_for_check: return False
    return True

def col_check(N, numbers):
  sum_for_check = get_sum_for_check(N)
  for i in range(0, N):
    col_sum = 0
    for j in range(0, N * N, N):
      col_sum += numbers[i + j]

    if not col_sum is sum_for_check: return False
    return True

def grid_check(N, numbers):
  sum_for_check = get_sum_for_check(N)
  sqrt_N = int(math.sqrt(N))
  for grid_x in range(0, sqrt_N):
    for grid_y in range(0, sqrt_N):
      grid_sum = 0
      for i in range(0, sqrt_N):
        for j in range(0, sqrt_N):
          grid_sum += numbers[N * i + j + grid_y * (N * sqrt_N) + grid_x * sqrt_N]

      if not grid_sum is sum_for_check: return False
      return True


for line in open(sys.argv[1], 'r'):
  line = line.rstrip('\n').split(';')
  N = int(line[0])
  numbers = map(int, line[1].replace(',', ' ').split(' '))

  if row_check(N, numbers) and col_check(N, numbers) and grid_check(N, numbers): 
    print True
  else:
    print False
